### PR TITLE
swarm/pss: Message handler refactor

### PR DESCRIPTION
**Pull request by nolash**
This is the first in a series of PR providing the following enhancements to PSS:

## Send fully addressed messages to neighbourhoods

The possibility of sending a message with full `PssAddress` to a neighbourhood of nodes, by interactive with the Kademlia. The primary use case is chunk syncing, where the chunk will have full address but nodes that have responsibility for storing the chunk should still regard themselves as recipients of it.

## Mitigate external influence of resource expenditure

If receiving raw messages, currently all raw messages will be passed on to handling code, effectively passing the responsibility of limiting resource expenditure to the handlers. This can be risky, in case of implementations that do not protect against request flooding. **Whether to handle raw messages or not should therefore be explicitly set on the handler**.

This also pertains to symmetric decryptions, were all registered symmetric keys (for decryption) share the same LRU cache which is tried sequentially on the decrypt. **Instead, there should be symkey decrypt caches per topic.** 

## Symkey filter in message type

To reduce resource expenditure in symmetric decryption, an extra field in the message type will be introduced where up to 32 bytes of symmetric key hint can be added. The hint is a hash of the symmetric key, indexed on both sides. The symkey cache must be refactored to be able to take advantage of this hint. A zero-value in this field would indicate public key crypto.

This also clears the way for upgrading whisper support to v6, as one of the changes in whisper v6 is omission of the symkey nonce in the whisper envelope; with this field it's still possible to identify incoming symmetrically encrypted messages, to avoid a forced attempt of asymmetric decryption on every reception.

# Part 1 

_`protocols`, `handshakes`, `pss/client` and `notify` have all been edited to pass tests with the new `Handler` struct, but need to be revisited for implementation details. This will follow in a later PR_

- [x] - Change Handler from func to struct with necessary params
- [x] - Set raw message allowance per handler
- [x] - Add possible recipient logic to use kademlia depth
- [x] - Handler capabilities tests
- [x] - Prox destinations tests
- [x] - Make sender handle message if topic has prox handler and is possible recipient

--- 

The PR also replaces the `-v[v]` verbosity flags for tests with `--loglevel` flag that sets the loglevel explicitly, and default level to 2 (`log.Warn`).

### CODE REVIEW

File: swarm/pss/pss.go [hunk 1]
```diff
@@ -422,19 +467,27 @@ func (p *Pss) process(pssmsg *PssMsg) error {
...
 
-func (p *Pss) executeHandlers(topic Topic, payload []byte, from *PssAddress, asymmetric bool, keyid string) {
+func (p *Pss) executeHandlers(topic Topic, payload []byte, from *PssAddress, raw bool, prox bool, asymmetric bool, keyid string) {
 	handlers := p.getHandlers(topic)
 	peer := p2p.NewPeer(enode.ID{}, fmt.Sprintf("%x", from), []p2p.Cap{})
-	for f := range handlers {
-		err := (*f)(payload, peer, asymmetric, keyid)
+	for h := range handlers {
+		if !h.caps.raw && raw {
+			log.Warn("norawhandler")
```
Comment (nonsense) [inline swarm/pss/pss.go]: Why do we need a warning here, and what will it help us debug if it happens?

File: swarm/pss/pss.go [hunk 2]
```diff
@@ -422,19 +467,27 @@ func (p *Pss) process(pssmsg *PssMsg) error {
...
+			continue
+		}
+		if !h.caps.prox && prox {
+			log.Warn("noproxhandler")
```
Comment (nonsense) [inline swarm/pss/pss.go]: Why do we need a warning here, and what will it help us debug if it happens?

File: swarm/pss/pss.go [hunk 3]
```diff
@@ -313,30 +316,54 @@ func (p *Pss) PublicKey() *ecdsa.PublicKey {
...
+	if hndlr.caps == nil {
+		hndlr.caps = &handlerCaps{}
+	}
+	handlers[hndlr] = true
+	if _, ok := p.topicHandlerCaps[*topic]; !ok {
+		p.topicHandlerCaps[*topic] = &handlerCaps{}
 	}
-	handlers[&handler] = true
-	return func() { p.deregister(topic, &handler) }
+	if !p.topicHandlerCaps[*topic].raw && hndlr.caps.raw {
```
Comment (nonsense) [inline swarm/pss/pss.go]: This can be simplified, the first conditional is redundant.

File: swarm/pss/pss.go [hunk 4]
```diff
@@ -313,30 +316,54 @@ func (p *Pss) PublicKey() *ecdsa.PublicKey {
...
+		p.topicHandlerCaps[*topic].raw = true
+	}
+	if !p.topicHandlerCaps[*topic].prox && hndlr.caps.prox {
```
Comment (nonsense) [inline swarm/pss/pss.go]: This can be simplified, the first conditional is redundant.

File: swarm/pss/pss.go (hunk 1 above)
Comment (nolash) [reply swarm/pss/pss.go]: The intention is that we are warned if the `topicHandlerCaps` flag tells us we shouldn't have a raw/prox handler, but in fact there is one. This condition shouldn't exist.

File: swarm/pss/types.go [hunk 5]
```diff
@@ -159,9 +159,39 @@ func (msg *PssMsg) String() string {
...
+
+// NewHandler returns a new message handler
+func NewHandler(f HandlerFunc) *handler {
+	return &handler{
+		f:    f,
+		caps: &handlerCaps{},
+	}
+}
+
+// WithRaw is a chainable method that allows raw messages to be handled.
```
[the remark this answers is not in the document]
Comment (nolash) [reply swarm/pss/types.go]: Earlier we agreed that we should use chained constructors for setting operating modes on objects, and clients will need a way to set these properties.

File: swarm/pss/pss.go (hunk 3 above)
Comment (nolash) [reply swarm/pss/pss.go]: Well spotted, sir. Thanks.

**Pull request merged**